### PR TITLE
[DPP-84] Employ parallel unnesting instead of batching on PostgreSQL

### PR DESCRIPTION
**Pull request by stefanobaghino-da**
changelog_begin
[Integration Kit] When using a PostgreSQL-based index, leveraging native
parallel unnesting allows to more efficiently index new transactions.
changelog_end

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by stefanobaghino-da** [comment]
Indexing-only benchmarking shows an improvement of 5%.
End-to-end benchmarking seems to confirm the improvement.

**Comment by stefanobaghino-da** [comment]
@mziolekda I noticed you closed the ticket for this PR but did not approve it. This cannot be merged until the PR is approved.

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/EventsTablePostgresql.scala
Comment (SamirTalwar-DA) [inline ledger/participant-integration-api/src/main/scala/platform/store/dao/events/EventsTablePostgresql.scala:46]: You can use `Array.ofDim[string](capacity)` here. It's terser, but I think also a little more performant; I have a feeling `fill` will actively loop over the array and insert `null` into every element slot.

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/EventsTablePostgresql.scala
Comment (SamirTalwar-DA) [inline ledger/participant-integration-api/src/main/scala/platform/store/dao/events/EventsTablePostgresql.scala:45]: Perhaps `batchSize`, rather than `capacity`? The latter implies we may never need to reach it. For example, I can have a vector of _capacity_ 1024 but only ever put 100 items in it, and its _size_ will be 100.

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/EventsTablePostgresql.scala
Comment (SamirTalwar-DA) [inline ledger/participant-integration-api/src/main/scala/platform/store/dao/events/EventsTablePostgresql.scala:78]: Why not inline this?

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/EventsTablePostgresql.scala
Comment (SamirTalwar-DA) [inline ledger/participant-integration-api/src/main/scala/platform/store/dao/events/EventsTablePostgresql.scala:93]: (And this.)

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/EventsTablePostgresql.scala
Comment (SamirTalwar-DA) [inline ledger/participant-integration-api/src/main/scala/platform/store/dao/events/EventsTablePostgresql.scala:155]: This `map` (and the equivalent in `IntArrayToStatement`) looks expensive. Why not just make the array one of `java.lang.Boolean` to start with?

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/EventsTablePostgresql.scala
Comment (SamirTalwar-DA) [inline ledger/participant-integration-api/src/main/scala/platform/store/dao/events/EventsTablePostgresql.scala:53]: TIL that `null.asInstanceOf[Int]` is `0`. This surprised me, and I would argue it's very confusing behavior. `null.asInstanceOf[Boolean]` is `false`.

Together with the `v.map(Int.box)` below (and the equivalent for booleans), I think boxed integers and booleans would be much more sensible here, if you really want nulls. If you don't, then please use `0` or `false`.

**Review by SamirTalwar-DA — COMMENTED** [review]
I'm going to take your word that the SQL is equivalent, because I tried to figure out where the old logic was and gave up.

I have some comments on the new implementation though.

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/EventsTablePostgresql.scala
Comment (stefanobaghino-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/events/EventsTablePostgresql.scala:155]: Because Scala optimizes it unless I box explicitly, unfortunately.

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/EventsTablePostgresql.scala
Comment (stefanobaghino-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/events/EventsTablePostgresql.scala]: Addressed by https://github.com/digital-asset/daml/pull/8042/commits/bb952a2afc862305d03dcc8060566df5114fca2b

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/EventsTablePostgresql.scala
Comment (stefanobaghino-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/events/EventsTablePostgresql.scala]: Addressed by https://github.com/digital-asset/daml/pull/8042/commits/bb952a2afc862305d03dcc8060566df5114fca2b

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/EventsTablePostgresql.scala
Comment (stefanobaghino-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/events/EventsTablePostgresql.scala]: Addressed by https://github.com/digital-asset/daml/pull/8042/commits/bb952a2afc862305d03dcc8060566df5114fca2b

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/EventsTablePostgresql.scala
Comment (stefanobaghino-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/events/EventsTablePostgresql.scala]: Addressed by https://github.com/digital-asset/daml/pull/8042/commits/bb952a2afc862305d03dcc8060566df5114fca2b

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/EventsTablePostgresql.scala
Comment (stefanobaghino-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/events/EventsTablePostgresql.scala]: Addressed by https://github.com/digital-asset/daml/pull/8042/commits/bb952a2afc862305d03dcc8060566df5114fca2b

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/EventsTablePostgresql.scala
Comment (SamirTalwar-DA) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/events/EventsTablePostgresql.scala]: Are you sure?

```scala
scala> val scalaBools = Array.ofDim[Boolean](5)
scalaBools: Array[Boolean] = Array(false, false, false, false, false)

scala> scalaBools(0)
res5: Boolean = false

scala> scalaBools(0).getClass
res6: Class[Boolean] = boolean

scala> val javaBools = Array.ofDim[java.lang.Boolean](5)
bools: Array[Boolean] = Array(null, null, null, null, null)

scala> javaBools(0)
res3: Boolean = null

scala> javaBools(0).getClass
java.lang.NullPointerException
  ... 28 elided
```

Comment (SamirTalwar-DA) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/events/EventsTablePostgresql.scala]: Ah, I see you _did_ change it. Cool. ^_^

**Review by SamirTalwar-DA — APPROVED** [review]
I approve this message. :+1:

**Pull request merged**
